subprojects/enterprise/src/main/java/org/gradle/internal/tasks/impl/DefaultTaskPropertiesService.java

```diff
@@ -0,0 +1,101 @@
+import org.gradle.api.internal.tasks.properties.OutputFilePropertyType;
+import org.gradle.api.internal.tasks.properties.PropertyValue;
+import org.gradle.api.internal.tasks.properties.PropertyVisitor;
+import org.gradle.api.internal.tasks.properties.PropertyWalker;
+import org.gradle.api.tasks.FileNormalizer;
+import org.gradle.internal.file.TreeType;
+import org.gradle.internal.fingerprint.DirectorySensitivity;
+import org.gradle.internal.fingerprint.LineEndingSensitivity;
+import org.gradle.internal.tasks.InputFileProperty;
+import org.gradle.internal.tasks.OutputFileProperty;
+import org.gradle.internal.tasks.TaskProperties;
+import org.gradle.internal.tasks.TaskPropertiesService;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.io.File;
+import java.util.Set;
+
+import static java.util.Collections.emptySet;
+
+public class DefaultTaskPropertiesService implements TaskPropertiesService {
+
+    private final PropertyWalker propertyWalker;
+    private final FileCollectionFactory fileCollectionFactory;
+
+    @Inject
+    public DefaultTaskPropertiesService(PropertyWalker propertyWalker, FileCollectionFactory fileCollectionFactory) {
+        this.propertyWalker = propertyWalker;
+        this.fileCollectionFactory = fileCollectionFactory;
+    }
+
+    @Override
+    public TaskProperties collectProperties(Task task) {
+        ImmutableSet.Builder<InputFileProperty> inputFiles = ImmutableSet.builder();
+        ImmutableSet.Builder<OutputFileProperty> outputFiles = ImmutableSet.builder();
+        TaskPropertyUtils.visitProperties(propertyWalker, (TaskInternal) task, new PropertyVisitor.Adapter() {
+            @Override
+            public void visitInputFileProperty(
+                String propertyName,
+                boolean optional,
+                boolean skipWhenEmpty,
+                DirectorySensitivity directorySensitivity,
+                LineEndingSensitivity lineEndingSensitivity,
+                boolean incremental,
+                @Nullable Class<? extends FileNormalizer> fileNormalizer,
+                PropertyValue value,
+                InputFilePropertyType filePropertyType,
+                ContentTracking contentTracking
+            ) {
+                Set<File> files = resolveLeniently(value);
+                inputFiles.add(new DefaultInputFileProperty(propertyName, files));
+            }
+
+            @Override
+            public void visitOutputFileProperty(
+                String propertyName,
+                boolean optional,
+                ContentTracking contentTracking,
+                PropertyValue value,
+                OutputFilePropertyType filePropertyType
+            ) {
+                Set<File> files = resolveLeniently(value);
+                OutputFileProperty.TreeType treeType = filePropertyType.getOutputType() == TreeType.DIRECTORY
+                    ? OutputFileProperty.TreeType.DIRECTORY
+                    : OutputFileProperty.TreeType.FILE;
```

Comment:
Does it really makes sense that me map Gradle's enum with values `FILE, DIRECTORY, FILES, DIRECTORIES` to our restricted enum `FILE, DIRECTORY`?

Reply:
`org.gradle.internal.file.TreeType` only has `FILE` and `DIRECTORY`

Reply:
Whoops!